lib/roo_on_rails/context_logging.rb

```diff
@@ -0,0 +1,97 @@
+require 'active_support/core_ext/module/delegation'
+require 'active_support/core_ext/object/blank'
+require 'active_support/version'
+require 'logger'
+require 'roo_on_rails/logfmt'
+
+module RooOnRails
+  # Wraps any standard logger to provide context, similar to `ActiveSupport::TaggedLogging`
+  # but with key/value pairs that are appended to the end of the text.
+  #
+  #   logger = RooOnRails::ContextLogging.new(Logger.new($stdout))
+  #   logger.with(a: 1, b: 2) { logger.info 'Stuff' }                   # Logs "Stuff -- a=1 b=2"
+  #   logger.with(a: 1) { logger.with(b: 2) { logger.info('Stuff') } }  # Logs "Stuff -- a=1 b=2"
+  #
+  # The above methods persist the context in thread local storage so it will be attached to
+  # any logs made within the scope of the block, even in called methods. However, if your
+  # context only applies to the current log then you can chain off the `with` method.
+  #
+  #   logger.with(a: 1, b: 2).info('Stuff')                   # Logs "Stuff -- a=1 b=2"
+  #   logger.with(a: 1) { logger.with(b: 2).info('Stuff')  }  # Logs "Stuff -- a=1 b=2"
+  module ContextLogging
+    module Formatter
+      def call(severity, timestamp, progname, msg)
+        super(severity, timestamp, progname, "#{msg}#{context_text}")
+      end
+
+      def with(**context)
+        current_context.push(context)
+        yield self
+      ensure
+        current_context.pop
+      end
+
+      def clear_context!
+        current_context.clear
+      end
+
+      private
+
+      def current_context
+        # We use our object ID here to avoid conflicting with other instances
+        thread_key = @thread_key ||= "roo_on_rails:logging_context:#{object_id}".freeze
+        Thread.current[thread_key] ||= []
```

Comment:
@danielcooper: the problem with the "middleware" approach is you still need to deal with e.g. Sidekiq. It also effectively precludes using async/threading in controllers.

A "typical" solution if we want to support logging continuity with e.g. promises is to perform context-passing — this is what [golang](https://blog.golang.org/context) favours for instance. Feels overkill for now.

We _do_ need to address your point though: as is, the code will break with servers that reuse threads (e.g. Thin, as per `request_store`'s example), because the context might not be reset properly on some exceptions. The `ensure` block need to do [this dance](https://github.com/mperham/connection_pool/blob/master/lib/connection_pool.rb#L59-L70), I believe.